src/compiler/checker.ts

```diff
@@ -4576,18 +4622,22 @@ namespace ts {
                 }
 
                 if (reportErrors) {
-                    headMessage = headMessage || Diagnostics.Type_0_is_not_assignable_to_type_1;
-                    let sourceType = typeToString(source);
-                    let targetType = typeToString(target);
-                    if (sourceType === targetType) {
-                        sourceType = typeToString(source, /*enclosingDeclaration*/ undefined, TypeFormatFlags.UseFullyQualifiedType);
-                        targetType = typeToString(target, /*enclosingDeclaration*/ undefined, TypeFormatFlags.UseFullyQualifiedType);
-                    }
-                    reportError(headMessage, sourceType, targetType);
+                    reportRelationError(headMessage, source, target);
                 }
                 return Ternary.False;
             }
 
+            function hasExcessProperties(source: FreshObjectLiteralType, target: Type, reportErrors: boolean): boolean {
+                for (let prop of getPropertiesOfObjectType(source)) {
+                    if (!isKnownProperty(target, prop.name)) {
+                        if (reportErrors) {
+                            reportError(Diagnostics.Property_0_does_not_exist_on_type_1, symbolToString(prop), typeToString(target));

```

Comment:
Given that we already have the specific codepath and context here to give a better error message I'd really like to see something more specific and distinct from a normal assignability error. I fully expect to see bug reports/Stack Overflow questions on this feature if the error simply looks like any other assignability error that changes when you add/remove temps in your chain of assignments. Minimally, a unique enough message that you could put in a search engine and get a distinct StackOverflow answer. Ideally, a message that actually communicates why assignments of this form differ from a regular assignment to a sufficient degree that you don't need to switch to StackOverflow to understand the error you just made.